Occasionally weird errors like 'Unable to determine array type, found unknown header byte 0xc4 at reader offset 59'

## Description
I am currently playing with spring-zeebe to support testing in spring environments based on zeebe-process-test 1.3.0-alpha3. I occasionally get the below exception on an assertion of process completion `assertThat(processInstance).isCompleted()`. Could it have to do with multi-threading, that some thread is still executing jobs and I do an assertion at the same time?

```
java.lang.RuntimeException: Could not deserialize object [JobBatchRecord]. Deserialization stuck at offset 60 of length 201

	at io.camunda.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:38)
	at io.camunda.zeebe.logstreams.impl.log.LoggedEventImpl.readValue(LoggedEventImpl.java:135)
	at io.camunda.zeebe.process.test.testengine.RecordStreamSourceImpl.mapToRecord(RecordStreamSourceImpl.java:169)
	at io.camunda.zeebe.process.test.testengine.RecordStreamSourceImpl.updateWithNewRecords(RecordStreamSourceImpl.java:148)
	at io.camunda.zeebe.process.test.testengine.RecordStreamSourceImpl.records(RecordStreamSourceImpl.java:46)
	at io.camunda.zeebe.process.test.testengine.RecordStreamSourceImpl.recordsOfValueType(RecordStreamSourceImpl.java:53)
	at io.camunda.zeebe.process.test.testengine.RecordStreamSourceImpl.processInstanceRecords(RecordStreamSourceImpl.java:60)
	at io.camunda.zeebe.process.test.filters.StreamFilter.processInstance(StreamFilter.java:9)
	at io.camunda.zeebe.process.test.assertions.ProcessInstanceAssert.isProcessInstanceCompleted(ProcessInstanceAssert.java:106)
	at io.camunda.zeebe.process.test.assertions.ProcessInstanceAssert.isCompleted(ProcessInstanceAssert.java:82)
	at io.berndruecker.playground.zeebe.tests.SimpleSpringBootTest.lambda$waitForCompletion$0(SimpleSpringBootTest.java:77)
	at org.testcontainers.shaded.org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
	at org.testcontainers.shaded.org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:234)
	at org.testcontainers.shaded.org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:221)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.lang.RuntimeException: Could not read property 'variables'
	at io.camunda.zeebe.msgpack.value.ObjectValue.read(ObjectValue.java:112)
	at io.camunda.zeebe.msgpack.UnpackedObject.wrap(UnpackedObject.java:32)
	... 17 more
Caused by: io.camunda.zeebe.msgpack.spec.MsgpackReaderException: Unable to determine array type, found unknown header byte 0xc4 at reader offset 59
	at io.camunda.zeebe.msgpack.spec.MsgPackReader.exceptionOnUnknownHeader(MsgPackReader.java:474)
	at io.camunda.zeebe.msgpack.spec.MsgPackReader.readArrayHeader(MsgPackReader.java:107)
	at io.camunda.zeebe.msgpack.value.ArrayValue.read(ArrayValue.java:94)
	at io.camunda.zeebe.msgpack.property.BaseProperty.read(BaseProperty.java:75)
	at io.camunda.zeebe.msgpack.value.ObjectValue.read(ObjectValue.java:110)
	... 18 more
```

@remcowesterhoud maybe?

--- a/src/main/java/io/camunda/zeebe/process/test/testengine/RecordStreamSourceImpl.java
+++ b/src/main/java/io/camunda/zeebe/process/test/testengine/RecordStreamSourceImpl.java
@@ -33,8 +33,8 @@ public class RecordStreamSourceImpl implements RecordStreamSource {
 
   private final LogStreamReader logStreamReader;
   private final int partitionId;
-  private List<Record<?>> records = new ArrayList<>();
-  private long lastPosition = -1L;
+  private final List<Record<?>> records = new ArrayList<>();
+  private volatile long lastPosition = -1L;
 
   public RecordStreamSourceImpl(final LogStreamReader logStreamReader, final int partitionId) {
     this.logStreamReader = logStreamReader;
@@ -137,17 +137,19 @@ public class RecordStreamSourceImpl implements RecordStreamSource {
   }
 
   private void updateWithNewRecords() {
-    if (lastPosition < 0) {
-      logStreamReader.seekToFirstEvent();
-    } else {
-      logStreamReader.seekToNextEvent(lastPosition);
-    }
+    synchronized (logStreamReader) {
+      if (lastPosition < 0) {
+        logStreamReader.seekToFirstEvent();
+      } else {
+        logStreamReader.seekToNextEvent(lastPosition);
+      }
 
-    while (logStreamReader.hasNext()) {
-      final LoggedEvent event = logStreamReader.next();
-      final CopiedRecord<UnifiedRecordValue> record = mapToRecord(event);
-      records.add(record);
-      lastPosition = event.getPosition();
+      while (logStreamReader.hasNext()) {
+        final LoggedEvent event = logStreamReader.next();
+        final CopiedRecord<UnifiedRecordValue> record = mapToRecord(event);
+        records.add(record);
+        lastPosition = event.getPosition();
+      }
     }
   }
 

--- a/src/test/java/io/camunda/zeebe/process/test/multithread/MultiThreadTest.java
+++ b/src/test/java/io/camunda/zeebe/process/test/multithread/MultiThreadTest.java
@@ -1,0 +1,69 @@
+package io.camunda.zeebe.process.test.multithread;
+
+import static io.camunda.zeebe.process.test.assertions.BpmnAssert.assertThat;
+import static io.camunda.zeebe.process.test.util.Utilities.deployProcess;
+import static io.camunda.zeebe.process.test.util.Utilities.startProcessInstance;
+
+import io.camunda.zeebe.client.ZeebeClient;
+import io.camunda.zeebe.client.api.response.ProcessInstanceEvent;
+import io.camunda.zeebe.process.test.RecordStreamSourceStore;
+import io.camunda.zeebe.process.test.extensions.ZeebeProcessTest;
+import io.camunda.zeebe.process.test.testengine.InMemoryEngine;
+import io.camunda.zeebe.process.test.testengine.RecordStreamSource;
+import io.camunda.zeebe.process.test.util.Utilities.ProcessPackStartEndEvent;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import org.apache.commons.lang3.exception.ExceptionUtils;
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+@ZeebeProcessTest
+public class MultiThreadTest {
+
+  private InMemoryEngine engine;
+  private ZeebeClient client;
+  private RecordStreamSource recordStreamSource;
+
+  @Test
+  public void testMultiThreadingThrowsNoExceptions() throws InterruptedException {
+    final ExecutorService executorService = Executors.newFixedThreadPool(5);
+
+    final List<Future<Boolean>> futures =
+        executorService.invokeAll(
+            List.of(
+                new ProcessRunner(),
+                new ProcessRunner(),
+                new ProcessRunner(),
+                new ProcessRunner(),
+                new ProcessRunner()));
+
+    for (final Future<Boolean> future : futures) {
+      try {
+        Assertions.assertThat(future.get()).isTrue();
+      } catch (ExecutionException ex) {
+        Assertions.fail("Future completed exceptionally: %s", ExceptionUtils.getStackTrace(ex));
+      }
+    }
+  }
+
+  private class ProcessRunner implements Callable<Boolean> {
+
+    @Override
+    public Boolean call() {
+      RecordStreamSourceStore.init(recordStreamSource);
+
+      deployProcess(client, ProcessPackStartEndEvent.RESOURCE_NAME);
+      final ProcessInstanceEvent instanceEvent =
+          startProcessInstance(engine, client, ProcessPackStartEndEvent.PROCESS_ID);
+      engine.waitForIdleState();
+
+      assertThat(instanceEvent).isCompleted();
+      RecordStreamSourceStore.reset();
+      return true;
+    }
+  }
+}

--- a/src/test/java/io/camunda/zeebe/process/test/util/Utilities.java
+++ b/src/test/java/io/camunda/zeebe/process/test/util/Utilities.java
@@ -62,8 +62,13 @@ public class Utilities {
     public static final String RESOURCE_NAME = "call-activity.bpmn";
     public static final String PROCESS_ID = "call-activity";
     public static final String CALL_ACTIVITY_ID = "callactivity";
-    public static final String CALLED_RESOURCE_NAME = "start-end.bpmn";
-    public static final String CALLED_PROCESS_ID = "start-end";
+    public static final String CALLED_RESOURCE_NAME = ProcessPackStartEndEvent.RESOURCE_NAME;
+    public static final String CALLED_PROCESS_ID = ProcessPackStartEndEvent.PROCESS_ID;
+  }
+
+  public static final class ProcessPackStartEndEvent {
+    public static final String RESOURCE_NAME = "start-end.bpmn";
+    public static final String PROCESS_ID = "start-end";
   }
 
   public static DeploymentEvent deployProcess(final ZeebeClient client, final String process) {

